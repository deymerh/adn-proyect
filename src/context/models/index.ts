export interface Cuenta {
  idCuenta: number;
  tipoCuenta: string;
  estado: number;
  valorDisponible: number;
}
export interface CuentasUserArray {
  CuentasUser: Array<Cuenta>;
}
