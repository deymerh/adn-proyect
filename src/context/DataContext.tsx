import { createContext } from 'react';
import { DepartamentosFundacion } from '../pages/IngresoDonaciones/models/DepartamentosFundacion';

export type GlobalContent = {
  globalState: DepartamentosFundacion;
  updateGlobalState: (p: DepartamentosFundacion) => void;
};

export const DataContext = createContext<GlobalContent>({
  globalState: {
    ninos: 0,
    construccion: 0,
    caridad: 0,
    dineroTotal: 0,
    personasDonantes: [],
  },
  updateGlobalState: (p: DepartamentosFundacion) => {
    return {};
  },
});
