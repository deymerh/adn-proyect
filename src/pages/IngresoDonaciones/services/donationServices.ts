import { DepartamentosFundacion } from '../models/DepartamentosFundacion';
import { UserDonator } from '../models/userDonator';

const mayorDeEdad = 18;
const porcentajeNinos = 0.40;
const porcentajeCaridad = 0.40;
const porcentajeContruccion = 0.20;
// eslint-disable-next-line
export const manejadorDonaciones = (
  usuarioDonante: UserDonator,
  globalState: DepartamentosFundacion,
  setInitialSate: (p: DepartamentosFundacion) => void,
) => {
  if (usuarioDonante.edad < mayorDeEdad) {
    const nuevoEstado: DepartamentosFundacion = {
      ...globalState,
      dineroTotal:
        Number(usuarioDonante.cantidadDonada) + Number(globalState.dineroTotal),
      ninos: Number(globalState.ninos) + Number(usuarioDonante.cantidadDonada),
      personasDonantes: [...globalState.personasDonantes, usuarioDonante],
    };
    setInitialSate(nuevoEstado);
  } else {
    const nuevoEstado: DepartamentosFundacion = {
      dineroTotal:
        Number(usuarioDonante.cantidadDonada) + Number(globalState.dineroTotal),
      caridad:
        Number(globalState.caridad) +
        Number(usuarioDonante.cantidadDonada * porcentajeCaridad),
      ninos:
        Number(globalState.ninos) + Number(usuarioDonante.cantidadDonada * porcentajeNinos),
      construccion:
        Number(globalState.construccion) +
        Number(usuarioDonante.cantidadDonada * porcentajeContruccion),
      personasDonantes: [...globalState.personasDonantes, usuarioDonante],
    };
    setInitialSate(nuevoEstado);
  }
};
