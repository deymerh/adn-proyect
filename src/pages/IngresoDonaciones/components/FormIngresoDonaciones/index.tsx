import React, { ChangeEvent, useState, useContext } from 'react';
import { Typography, Grid, Button } from '@material-ui/core';
import { Link } from 'react-router-dom';
import { DataContext } from '../../../../context/DataContext';
import { useStyles } from './styles';
import { Donante } from '../../models/Donante';
import { InputText } from '../Input/index';
import * as donationServices from '../../services/donationServices';

export const IngresoDonaciones: React.FC = () => {
  const classes = useStyles();
  const { globalState, updateGlobalState } = useContext(DataContext);
  const [usuarioDonante, setUsuarioDonante] = useState<Donante>({
    nombre: 'Deymer',
    id: 73984410,
    cantidadDonada: 10000,
    edad: 17,
    fecha: '2021-07-30',
  });
  const handleInputChange = ({ target }: ChangeEvent<HTMLInputElement>) => {
    setUsuarioDonante({
      ...usuarioDonante,
      [target.name]: target.value,
    });
  };
  const handleSubmit = () => {
    donationServices.manejadorDonaciones(
      usuarioDonante,
      globalState,
      updateGlobalState,
    );
  };
  return (
    <Grid container className={classes.container}>
      <Typography variant="h5" color="textSecondary">
        Ingresar donaci√≥n
      </Typography>
      <Grid justifyContent="center" container item xs={10} md={6}>
        <Grid item md={6} xs={12}>
          <InputText
            value={usuarioDonante.nombre}
            name="nombre"
            onChange={handleInputChange}
          />
          <InputText
            value={usuarioDonante.id}
            name="id"
            type="number"
            onChange={handleInputChange}
          />
          <InputText
            type="number"
            name="cantidadDonada"
            onChange={handleInputChange}
            value={usuarioDonante.cantidadDonada}
          />
          <InputText
            type="number"
            name="edad"
            onChange={handleInputChange}
            value={usuarioDonante.edad}
          />
          <InputText
            type="date"
            name="fecha"
            onChange={handleInputChange}
            value={usuarioDonante.fecha}
          />
          <Button
            className={classes.btn}
            fullWidth
            variant="contained"
            color="primary"
            size="small"
            onClick={handleSubmit}
          >
            Guardar
          </Button>
        </Grid>
      </Grid>
      {globalState.personasDonantes.length > 0 && (
        <Button size="small">
          <Link to="/donaciones">Ver donaciones</Link>
        </Button>
      )}
    </Grid>
  );
};
