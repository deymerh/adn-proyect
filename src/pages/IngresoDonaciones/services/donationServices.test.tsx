import '@testing-library/jest-dom';
import React from 'react';
import { act } from '@testing-library/react-hooks';
import { mount } from 'enzyme';
import { manejadorDonaciones } from './donationServices';
import { IngresoDonaciones } from '../components/FormIngresoDonaciones/index';
import { DataContext } from '../../../context/DataContext';

jest.mock('./donationServices');
jest.mock('../../../context/DataContext');

describe('Pruebas en la función ´manejadorDonaciones´ ', () => {
  const updateGlobalState = jest.fn();
  const globalState = {
    ninos: 0,
    construccion: 0,
    caridad: 0,
    dineroTotal: 0,
    personasDonantes: [],
  };
  const wrapper = mount(
    <DataContext.Provider value={{ globalState, updateGlobalState }}>
      <IngresoDonaciones />
    </DataContext.Provider>,
  );
  test('debe crear el snapshot de con el ´DataContext´ ', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('`updateGlobalState` debe ser llamado con el objeto usuario', () => {
    // Arrange
    const usuario = {
      nombre: 'Deymer',
      id: 708098,
      cantidadDonada: 10000,
      edad: 17,
      fecha: '2021-07-19',
    };
    // Act
    act(() => {
      manejadorDonaciones(usuario, globalState, updateGlobalState);
      updateGlobalState(usuario);
    });
    // Assert
    expect(manejadorDonaciones).toBeCalledTimes(1);
    expect(updateGlobalState).toHaveBeenCalledWith(usuario);
  });
});
