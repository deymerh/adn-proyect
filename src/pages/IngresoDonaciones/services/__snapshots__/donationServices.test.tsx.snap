// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en la función ´manejadorDonaciones´  debe crear el snapshot de con el ´DataContext´  1`] = `ReactWrapper {}`;
